File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Senegal.m
//
//  UIImage+Senegal.m
//  CountriesFlags
//

#import "Countries/UIImage+Senegal.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (Senegal)

+ (UIImage *)flagSenegalWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:133.0f/255.0f
                                           blue:63.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:253.0f/255.0f
                                           green:239.0f/255.0f
                                            blue:66.0f/255.0f
                                           alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:227.0f/255.0f
                                        green:27.0f/255.0f
                                         blue:35.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    UIRectFill(CGRectMake(size.width / 3.0, 0.0f, size.width / 3.0, size.height));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(size.width / 3.0 * 2.0, 0.0f, size.width / 3.0, size.height));
    
    [colorGreen setFill];
    [[UIBezierPath starAtPoint:CGSizeMidPoint(size)
                        radius:size.height * 5.0f / 24.0f
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Senegal.swift
//
//  UIImage+Senegal.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Senegal "wikipedia"
     
     Generates the flag of Senegal.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Senegal.
     */
    internal static func flagSenegal(with size: CGSize) -> UIImage {
        let colorGreen = UIColor(red: 0.0,
                                 green: 133.0 / 255.0,
                                 blue: 63.0 / 255.0,
                                 alpha: 1.0)
        let colorYellow = UIColor(red: 253.0 / 255.0,
                                  green: 239.0 / 255.0,
                                  blue: 66.0 / 255.0,
                                  alpha: 1.0)
        let colorRed = UIColor(red: 227.0 / 255.0,
                               green: 27.0 / 255.0,
                               blue: 35.0 / 255.0,
                               alpha: 1.0)
        let lineSize = CGSize(width: size.width / 3.0, height: size.height)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 3.0, y: 0.0), size: lineSize))

        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 3.0 * 2.0, y: 0.0), size: lineSize))

        colorGreen.setFill()
        UIBezierPath.star(at: size.midPoint, radius: size.height * 5.0 / 24.0).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
